contracts/extensions/VotingReputation.sol

```diff
@@ -0,0 +1,817 @@
+      repPenalty = (winShare < WAD) ? sub(stake, wmul(winShare, stake)) : 0;
+
+    // Your side fully staked, receive 10% (proportional) of loser's stake
+    } else if (poll.stakes[_vote] == getRequiredStake(_pollId)) {
+      uint256 stakePercent = wdiv(stake, poll.stakes[_vote]);
+      uint256 totalPenalty = wmul(poll.stakes[flip(_vote)], WAD / 10);
+      stakerReward = add(rewardStake, wmul(stakePercent, totalPenalty));
+
+    // Opponent's side fully staked, pay 10% penalty
+    } else if (poll.stakes[flip(_vote)] == getRequiredStake(_pollId)) {
+      stakerReward = wmul(rewardStake, (WAD / 10) * 9);
+      repPenalty = stake / 10;
+
+    // Neither side fully staked, no reward or penalty
+    } else {
+      stakerReward = rewardStake;
+    }
+
+    tokenLocking.transfer(token, stakerReward, _user, true);
+
+    if (repPenalty > 0) {
+      colony.emitDomainReputationPenalty(
+        _permissionDomainId,
+        _childSkillIndex,
+        poll.domainId,
+        _user,
+        -int256(repPenalty)
+      );
+    }
+
+    emit PollRewardClaimed(_pollId, _user, _vote, stakerReward);
+  }
+
+  // Public view functions
+
+  function getPollCount() public view returns (uint256) {
+    return pollCount;
+  }
+
+  function getPoll(uint256 _pollId) public view returns (Poll memory poll) {
+    poll = polls[_pollId];
+  }
+
+  function getStake(uint256 _pollId, address _staker, uint256 _vote) public view returns (uint256) {
+    return stakes[_pollId][_staker][_vote];
+  }
+
+  function getExpenditurePollCount(bytes32 _structHash) public view returns (uint256) {
+    return expenditurePollCounts[_structHash];
+  }
+
+  function getExpenditurePastPoll(bytes32 _slotHash) public view returns (uint256) {
+    return expenditurePastPolls[_slotHash];
+  }
+
+  function getPollState(uint256 _pollId) public view returns (PollState) {
+    Poll storage poll = polls[_pollId];
+    uint256 requiredStake = getRequiredStake(_pollId);
+
+    // If executed, we're done
+    if (poll.executed) {
+      return PollState.Executed;
+
+    // Not fully staked
+    } else if (
+      poll.stakes[YAY] < requiredStake ||
+      poll.stakes[NAY] < requiredStake
+    ) {
+      // Are we still staking?
+      if (now < poll.events[CREATE] + stakePeriod && poll.events[STAKE_END] == 0) {
+        return PollState.Staking;
+      // If not, did the YAY side stake?
+      } else if (poll.stakes[YAY] == requiredStake) {
+        return PollState.Executable;
+      // If not, was there a prior vote we can fall back on?
+      } else if (poll.votes[NAY] > 0 || poll.votes[YAY] > 0) {
+        return PollState.Executable;
+      // Otherwise, the poll failed
+      } else {
+        return PollState.Failed;
+      }
+
+    // Do we need to keep waiting?
+    } else if (now < poll.events[CREATE] + stakePeriod && poll.events[STAKE_END] == 0) {
+      return PollState.Staking;
+
+    // Fully staked, go to a vote
+    } else {
+
+      if (now < min(
```

Comment:
Heh, I guess this is what you had in mind when you were complaining about on our call about the code getting worse as things went on and new features are requested.

How about we make the poll.events array elements be of type `uint64` (or similar), and set them all when the poll is created (the type changing being so that it only use one storage slot)? So something like

```
    polls[pollCount].events[CREATE] = now;
    polls[pollCount].events[STAKE_END] = now + stakePeriod;
    polls[pollCount].events[SUBMIT_END] = polls[pollCount].events[STAKE_END] + submitPeriod;
    polls[pollCount].events[REVEAL_END] = polls[pollCount].events[SUBMIT_END] + revealPeriod;
```

If we trigger an 'early' end to something, we update all subsequent ending timestamps where we currently set the timestamps from 0. Then, this block of code just becomes

```
    } else if (now < poll.events[SUBMIT_END]){
      return PollState.Submit;
    } else if (now < poll.events[REVEAL_END]){
      return PollState.Reveal;
    } else if (now < poll.events[REVEAL_END] + escalationPeriod){
      return PollState.Closed;
    } else {
      return PollState.Executable;
    }
```
With the type change, there shouldn't be too much of a gas overhead, as we were already always setting `poll.events[CREATE]` (which maybe could be dropped, in this scenario?). We could even drop to assembly and set the slot in one go and it'd be very close to as cheap, storage wise. And this code would be a lot easier to read / maintain.

Reply:
I'll take it :)

Reply:
Actually, we probably save a chunk of gas if we avoid needless SLOAD`ing, so the setting code is probably better looking something like:

```
    uint256 time = now;
    polls[pollCount].events[CREATE] = time;
    time += stakePeriod;
    polls[pollCount].events[STAKE_END] = time;
    time += submitPeriod;
    polls[pollCount].events[SUBMIT_END] = time;
    time += revealPeriod
    polls[pollCount].events[REVEAL_END] = time;
```

if we don't go the whole hog and set the slot in one go.